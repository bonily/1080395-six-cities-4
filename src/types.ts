export interface City {
  name: string,
  coords: [number, number],
  zoom: number
};

export interface Offer {
  id: number,
  title: string,
  price: number,
  raiting: number,
  type: string,
  isInBookmark: boolean,
  isPremium: boolean,
  coords: [number, number],
  description: string,
  bedrooms: number,
  quests: {adults: number, kids: number},
  items:string[],
  host: {
    avatar: string,
    name: string,
    isSuper: boolean,
  },
  photos: string[],
  city: City
}

export interface Review {
    comment: string,
    user: {
      name: string,
      avatarUrl: string,
      rating: number
    },
    rating: number,
    date: string,
    id: number
}
