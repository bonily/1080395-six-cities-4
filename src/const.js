export const MAX_STAR_COUNT = 5;

export const FILTERS = [
  {
    name: `popular`,
    description: `Popular`
  },
  {
    name: `low`,
    description: `Price: low to high`
  },
  {
    name: `high`,
    description: `Price: high to low`
  },
  {
    name: `top`,
    description: `Top rated first`
  },
];

export const FILTER_NAMES = {
  POPULAR: `popular`,
  LOW: `low`,
  HIGH: `high`,
  TOP: `top`
};

export const ErrorTypes = {
  NETWORK: `Network Error`,
  BAD_REQUEST: 400,
};

export const AppRoute = {
  LOGIN: `/login`,
  FAVORITE: `/favorite`,
  OFFER: `/offer`,
  ROOT: `/`,
};

export const COMMENT_LENGTH = {
  MIN_LENGTH: 50,
  MAX_LENGTH: 300
};
