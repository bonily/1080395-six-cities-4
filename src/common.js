import {FILTER_NAMES} from "./const.js";


export const capitalize = (value) => {
  return value.charAt(0).toUpperCase() + value.slice(1);
};

export const split = (text) => {
  return text.split(`. `);
};

export const extend = (a, b) => {
  return Object.assign({}, a, b);
};

export const getUnicValues = (values) => {
  return values.filter((value, i) => values.indexOf(value) === i);
};

export const getOffersByCity = (city, offers) => {
  return offers[city];
};

export const getCitiesFromOffers = (offers) => {
  const cities = Object.keys(offers);
  const result = cities.map((city) => {
    return (
      offers[city][0].city
    );
  });

  return result;
};

export const getFilteredOffers = (offers, currentFilter) => {
  switch (currentFilter) {
    case FILTER_NAMES.LOW:
      return offers.sort((a, b) => a.price - b.price).slice();
    case FILTER_NAMES.HIGH:
      return offers.sort((a, b) => b.price - a.price).slice();
    case FILTER_NAMES.TOP:
      return offers.sort((a, b) => b.raiting - a.raiting).slice();
  }
  return offers.slice();
};

export const groupOffersByCity = (offers) =>
  offers.reduce((accumulator, offer) =>{
    const key = offer.city.name;

    if (accumulator[key] === void 0) {
      accumulator[key] = [];
    }
    accumulator[key].push(offer);

    return accumulator;
  }, {});

export const updateOffers = (offers, offer) => {

  const offerCity = offer.city.name;
  let currentOffers = getOffersByCity(offerCity, offers);
  console.log(currentOffers);
  const index = currentOffers.findIndex((currentOffer) => currentOffer.id === offer.id);
  console.log(index);
  currentOffers = [].concat(currentOffers.slice(0, index), offer, currentOffers.slice(index + 1));
  console.log(currentOffers);

  return extend(offers, {
    offerCity: currentOffers
  });
};
