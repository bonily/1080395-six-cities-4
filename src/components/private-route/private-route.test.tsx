import * as React from "react";
import * as renderer from "react-test-renderer";
import configureStore from "redux-mock-store";
import {Provider} from "react-redux";
import {PrivateRoute} from "./private-route";
import {noop} from "../../common";
import NameSpace from "../../reducer/name-space";
import Favorite from "../favorite/favorite";
import {Offer} from "../../types";
import {AppRoute} from "../../const";
import { Router, MemoryRouter } from "react-router-dom";
import history from "../../history.js";


const mockStore = configureStore([]);

const enum AuthorizationStatus {
  AUTH =`AUTH`,
  NO_AUTH =`NO_AUTH`
}
const userName = `oliver@gmail.com`;

const favoriteOffers: {[key: string]: Offer[]} = {
  Paris: [
    {
      id: 1,
      title: `Beautiful & luxurious apartment at great location`,
      description: `A quiet cozy and picturesque that hides behind a a river by the unique lightness of Amsterdam.The building is green and from 18th century. An independent House, strategically located between Rembrand Square and National Opera, but where the bustle of the city comes to rest in this alley flowery and colorful.`,
      price: 120,
      raiting: 4,
      bedrooms: 3,
      quests: {
        adults: 5,
        kids: 2
      },
      items: [`Wi-Fi`, `Heating`, `Kitchen`, `Cable TV`, `Baby seat`, `Dishwasher`, `Washing machine`],
      type: `apartment`,
      isInBookmark: false,
      isPremium: true,
      photos: [`img/1.png`, `img/2.png`],
      host: {
        avatar: `img/avatar-angelina.jpg`,
        name: `Angelina`,
        isSuper: true,
      },
      coords: [52.3909553943508, 4.85309666406198],
      city: {
        name: `Paris`,
        coords: [1, 2],
        zoom: 10
      }
    },
    {
      id: 2,
      title: `Wood and stone place`,
      description: ``,
      price: 80,
      raiting: 4,
      bedrooms: 1,
      quests: {
        adults: 2,
        kids: 1
      },
      items: [`Wi Fi`, `Heating`, `Cable TV`, `Dishwasher`],
      type: `room`,
      isInBookmark: true,
      isPremium: false,
      photos: [`img/1.png`, `img/2.png`],
      host: {
        avatar: `img/avatar-max.jpg`,
        name: `Masha`,
        isSuper: false,
      },
      coords: [52.369553943508, 4.85309666406198],
      city: {
        name: `Amsterdam`,
        coords: [1, 2],
        zoom: 10
      }
    },
    {
      id: 3,
      title: `Canal View Prinsengracht`,
      description: ``,
      price: 132,
      raiting: 4,
      bedrooms: 6,
      quests: {
        adults: 10,
        kids: 5
      },
      items: [`Wi Fi`, `Heating`, `Kitchen`, `Cable TV`, `Fridge`],
      type: `house`,
      isInBookmark: false,
      isPremium: false,
      photos: [`img/1.png`, `img/2.png`],
      host: {
        avatar: `img/avatar-olga.jpg`,
        name: `Olga`,
        isSuper: true,
      },
      coords: [52.3909553943508, 4.929309666406198],
      city: {
        name: `Cologne`,
        coords: [1, 2],
        zoom: 10
      }
    },
    {
      id: 4,
      title: `Nice, cozy, warm big bed apartment`,
      description: ``,
      price: 180,
      raiting: 4,
      bedrooms: 1,
      quests: {
        adults: 2,
        kids: 0
      },
      items: [`Wi-Fi`, `Cable TV`, `Coffee machine`, `Washing machine`],
      type: `hotel`,
      isInBookmark: false,
      isPremium: false,
      photos: [`img/1.png`, `img/2.png`],
      host: {
        avatar: `img/avatar-roma.jpg`,
        name: `Roman`,
        isSuper: false,
      },
      coords: [52.3809553943508, 4.939309666406198],
      city: {
        name: `Barcelona`,
        coords: [1, 2],
        zoom: 10
      }
    }]
};

const render = () => {
  return (
    <Favorite
      authorizationStatus = {AuthorizationStatus.AUTH}
      onOfferTitleClick = {noop}
      name = {userName}
      changeFavoriteStatus = {noop}
      favoriteOffers = {favoriteOffers}
    />
  );
};

describe(`PrivateRouterSnapTest`, () => {
  it(`PrivateRouter without authorization render LoginPage`, () => {
    const store = mockStore({
      [NameSpace.USER]: {
        authorizationStatus: AuthorizationStatus.AUTH,
      },
    })
    const tree = renderer
     .create(

      <Provider store={store}>
        <Router history = {history}>
          <PrivateRoute
            render = {render}
            path = {AppRoute.FAVORITE}
            authorizationStatus = {AuthorizationStatus.AUTH}
          />
        </Router>
      </Provider>
     )
     .toJSON();

     expect(tree).toMatchSnapshot();
  })
})
