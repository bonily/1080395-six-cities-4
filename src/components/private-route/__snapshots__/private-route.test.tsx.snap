// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PrivateRouterSnapTest PrivateRouter without authorization render LoginPage 1`] = `"        "`;
